File: src/main/java/kr/devport/api/domain/common/logging/LoggingContext.java
package kr.devport.api.domain.common.logging;

import org.slf4j.MDC;

import java.util.Map;
import java.util.UUID;

public final class LoggingContext {

    public static final String REQUEST_ID_HEADER = "X-Request-ID";
    public static final String REQUEST_ID_KEY = "requestId";

    private LoggingContext() {
    }

    public static String resolveRequestId(String requestIdHeader) {
        if (requestIdHeader != null && !requestIdHeader.isBlank()) {
            return requestIdHeader.strip();
        }
        return UUID.randomUUID().toString();
    }

    public static Runnable wrap(Runnable delegate) {
        Map<String, String> capturedContext = MDC.getCopyOfContextMap();
        return () -> {
            Map<String, String> previousContext = MDC.getCopyOfContextMap();
            try {
                if (capturedContext != null) {
                    MDC.setContextMap(capturedContext);
                } else {
                    MDC.clear();
                }
                delegate.run();
            } finally {
                if (previousContext != null) {
                    MDC.setContextMap(previousContext);
                } else {
                    MDC.clear();
                }
            }
        };
    }
}

File: support/logging/src/main/kotlin/kr/devport/api/domain/common/logging/LoggingContext.kt
package kr.devport.api.domain.common.logging

import org.slf4j.MDC
import java.util.UUID

object LoggingContext {
    const val REQUEST_ID_HEADER = "X-Request-ID"
    const val REQUEST_ID_KEY = "requestId"

    fun resolveRequestId(requestIdHeader: String?): String {
        if (!requestIdHeader.isNullOrBlank()) {
            return requestIdHeader.trim()
        }
        return UUID.randomUUID().toString()
    }

    fun wrap(delegate: Runnable): Runnable {
        val capturedContext: Map<String, String>? = MDC.getCopyOfContextMap()
        return Runnable {
            val previousContext: Map<String, String>? = MDC.getCopyOfContextMap()
            try {
                if (capturedContext != null) {
                    MDC.setContextMap(capturedContext)
                } else {
                    MDC.clear()
                }
                delegate.run()
            } finally {
                if (previousContext != null) {
                    MDC.setContextMap(previousContext)
                } else {
                    MDC.clear()
                }
            }
        }
    }
}
